### ISSUE TITLE
[ML Idea] Heirarchical Perceiver

### ISSUE THREAD

**Issue by jacobbieker** [openclimatefix/power_perceiver#14]
## Detailed Description
https://arxiv.org/abs/2202.10890 is a new Perceiver model that is supposed to scale better,etc. I've only skimmed it, but might be a nice improvement.

**Comment by JackKelly** [comment]
Thanks loads for the heads up! I hadn't seen that paper! Looks great.

**Comment by JackKelly** [comment]
I've only skim-read a few paragraphs so far (!) but, yeah, this could be really useful to allow us to include large geographical regions in the input! Nice! Thanks again.

**Comment by JackKelly** [comment]
I've started reading the paper. Some notes:

The paper is very focused on creating a one size fits all "AutoML" model. But in our case, we're happy to modify the architecture for our domain.

We could use the HiP idea by using one group per timestep (instead of trying to use a Perceiver as an RNN, where each cross-attend sees a timestep).

Advantages of using "one-HiP-group-per-timestep":
- The "higher levels" can see across all timesteps (forwards and backwards in time), which might be useful.

Disadvantages: In Perceiver, I liked the idea that cross attend _i_ can gets to query cross attend _i+1_. But maybe that's just a pretty idea, rather than something that's definitely skillful in practice!

The paper shows that learned position encodings kind of suck when trained purely on a classification task. In the paper, they use a (heavily) masked auto encoding to pre-train the learned position encoding. For us, we have a very natural dense prediction task (predicting a future satellite image) which may be sufficient to learn position encodings. That said, the paper's results on learned position encodings weren't _that_ amazing. So I think we're probably fine to stick with hand-crafted Fourier features for now.

One thing I'm a bit nervous about is the "fan out" (where they use, say, a hierarchy of 1-4-8 groups to decode the signal). Instead, I wonder if it might be nice to use HiP as the "history encoder", and then repeatedly cross-attend to the output of that "history encoder" to generate each timestep of prediction. Although I can't entirely put my finger on why I think that, right now! Maybe one thought is that I feel like every prediction timestep should have access to the entire latent bottleneck vector, and not just one group. And, because, if we want to predict days into the future, then we'll need, like, 48 output groups. Which sounds onorous. And each output group would only get a tiny input.

Right now, without thinking about it very deeply (and without having read the entire paper), I quite like the idea of using HiP as the "history encoder", and using repeated cross-attends to generate the predictions.

**Comment by JackKelly** [comment]
Or do the MetNet thing: run the entire model (encoder and decoder) for every prediction timestep. Ie the model always outputs exactly one timestep. And all the queries are conditioned on the target timestep

**Comment by JackKelly** [comment]
After sleeping on it, I really like the idea of kind of merging ideas from Hierarchical Perceiver (HiP) with MetNet.

To be specific: A model which:

* Predicts a single timestep at a time
* Has no expensive recurrent connections
* Tells the entire model the "forecast timestep": MetNet feeds the "forecast timestep" into the model as a channel to the CNN, if I remember correctly.

![image](https://user-images.githubusercontent.com/460756/158989735-d54772c9-86c2-4ed0-a1c4-a939d2176a1d.png)

[Link to Google Drawing of the diagram above](https://docs.google.com/drawings/d/1oryY7QeLGAKy7faKC8DCWK_RF91c0i03kMMcVUrIhiA/edit?usp=sharing).

**Comment by JackKelly** [comment]
@jacobbieker absolutely no rush but I'd be curious if you have any thoughts on the architecture sketched out above? :slightly_smiling_face:

**Comment by jacobbieker** [comment]
Yeah, I like the idea! One thing that MetNet-2 found was that instead of just giving the forecast timestep in the input, condition all the convolutional blocks with the timestep, I do it [here](https://github.com/openclimatefix/metnet/blob/349f2e47caa0564362b46d6adc547d29200e9bf1/metnet/models/metnet2.py#L114-L118) in our implementation, where a scale and bias are computed for each conv block based on teh forecast timestep and added to that layer, to ensure its strongly conditioned on the lead time. So I think something like that could help here too. But other than that, I think this looks great!

**Comment by JackKelly** [comment]
> MetNet-2 found was that instead of just giving the forecast timestep in the input, condition all the convolutional blocks with the timestep

ooh, that's really interesting, I had missed that detail! Very useful, thank you @jacobbieker!

**Comment by JackKelly** [comment]
After thinking about graph neural networks, I'm now thinking about an encoder / decoder model, more like this:

![image](https://user-images.githubusercontent.com/460756/161044166-f64f3510-2c22-4b39-9b5c-3437c19f5c9f.png)

(link to [Google Drawing](https://docs.google.com/drawings/d/1Mhx-0r76saNsyBL72Vvb42FHMYd9Zn2m8HORySsnM1g/edit)).

See issue #34 for more discussion of graph neural nets, and why the model above might be a nice compromise.
